src/attributes.rs

```diff
@@ -0,0 +1,235 @@
+use std::rc::Rc;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Attribute<'a> {
+    pub is_inner: bool,
+    pub content: Rc<AttributeValue<'a>>,
+}
+
+impl<'a> Attribute<'a> {
+    pub fn as_string(&self) -> String {
+        format!(
+            "#{}[{}]",
+            if self.is_inner { "!" } else { "" },
+            self.content.raw_value
+        )
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let raw_without_closing = raw
+            .trim()
+            .strip_suffix(']')
+            .expect("Attribute has to be closed by a square bracket.");
+
+        if let Some(raw_content) = raw_without_closing.strip_prefix("#[") {
+            Attribute {
+                is_inner: false,
+                content: Rc::new(AttributeValue::new(raw_content)),
+            }
+        } else if let Some(raw_content) = raw_without_closing.strip_prefix("#![") {
+            Attribute {
+                is_inner: true,
+                content: Rc::new(AttributeValue::new(raw_content)),
+            }
+        } else {
+            panic!("Attribute has to start either with `#[` or `#![`.");
```

Comment:
Consider what would happen if this `panic!()` got hit: we'd see this static string in the error message, and nothing else.

So the first thing we'd then do is edit the `panic!()` call to include additional context, like what the `raw` value that triggered this was. And then we'd try to replicate the crash to see the error with that additional context, so we can debug it further.

We can save a step here: just include all the necessary context immediately, without waiting for the panic to be hit in the real world first. This is a way to reduce future debugging time, without significant upfront cost, and is highly effective in my experience.

There are a few more instances of this kind of situation throughout this PR, and I'd encourage you to self-review it to practice noticing them as well as fixing them.

Reply:
You're definitely right, I'll change it!

Reply:
It should be better now :)